### PR TITLE
Activate OPL3 chips on-demand based on voice pressure

### PR DESCRIPTION
**Pull request by jpernst**
This turned out to be a little trickier than I thought, but the patch ended up being pretty short after all. In any case, this patch makes 2 modifications to the OPL driver:

First, chips with no active channels that have been outputting silence for at least half a second will be paused until they key-on again.
Second, DMX will attempt to allocate free voices from already active chips until all such voices are used, only then will it draw voices from the next available chip. With only 1 chip active, this is still identical to the vanilla behavior.

In combination, these changes allow the higher chips to go idle when not strictly needed, clawing back a significant amount of CPU time. Only extremely busy sections of a song will activate all of the chips, and they'll quickly fall idle again after the notes clear out.

### CODE REVIEW

**Comment by jpernst** [comment]
Just to add, there is one minor caveat to running multiple OPL chips asynchronously like this. Doing so causes their internal LFOs to drift out of sync with eachother, which can produce some minor audible beating artifacts under certain conditions. The circumstances needed for this to be noticeable are relatively rare, and the effect I think actually sounds kindof cool when it does happen, but it is nonetheless a difference from always running all of the chips.

If people do end up noticing that, a fix could be to offer a lighter but still accurate-enough emulation core like Opal that is fast enough to be always-on.

**Comment by fabiangreffrath** [comment]
This looks very clean, thank you!

> First, chips with no active channels that have been outputting silence for at least half a second will be paused until they key-on again.

Is this an arbitrary limit? 

Is it possible to skim through a midi lump and check how many voices will be necessary in advance?

**Comment by jpernst** [comment]
OPL chips have a hardware ADSR envelopes, which means that even when all channels are keyed off, it may still produce sound for many seconds afterwards as the notes fade away. Deactivating the chip too soon will cut them off awkwardly. Unfortunately, channels aren't guaranteed to fade away to pure zeros either; the oscillators are always running and continue to produce a very slight ripple indefinitely. The silence threshold check is to make sure it's been quiet enough for long enough to considered truly "off." There's nothing particularly special about the half-second timeframe, just that it's unlikely that the chip will produce any more audible sound if it's been quiet for that long.

As for pre-scanning songs, that would be possible, but probably more work that it's worth. It's been my observation that many songs have brief bursts of very-high polyphony that can push up to 4 chips or more, but only last for a few seconds, while most of the song is spent using 2 or even just 1 chip, so it makes sense to just add and remove chips on the fly.

**Comment by fabiangreffrath** [comment]
Thank you!

File: opl/opl_sdl.c [hunk 1]
```diff
@@ -23,6 +25,9 @@
 
 #define MAX_SOUND_SLICE_TIME 100 /* ms */
 
+#define OPL_SILENCE_THRESHOLD   36
```
Comment (fabiangreffrath) [inline opl/opl_sdl.c:28]: Could you add comments to the motivation of these values, please?

**Comment by rfomin** [comment]
> Just to add, there is one minor caveat to running multiple OPL chips asynchronously like this. Doing so causes their internal LFOs to drift out of sync with eachother, which can produce some minor audible beating artifacts under certain conditions. The circumstances needed for this to be noticeable are relatively rare, and the effect I think actually sounds kindof cool when it does happen, but it is nonetheless a difference from always running all of the chips.

Hmm, to be honest, I'm not sure this feature is worth such a trade-off. We use Nuked OPL and code from Chocolate Doom to maximize accuracy, if we allow such things, what's the point in heavy and accurate OPL emulation? I like the multi-chip feature as it is.

**Comment by jpernst** [comment]
I gave it some thought, and I have an idea for how to fix that issue. I think it's possible to resynchronize the LFOs when I reactivate the chips. I'll do a bit more investigation later today.

**Comment by jpernst** [comment]
Success. Turned out to be a rather easy fix after inspecting the Nuked internals. Any chip that wakes up after being idle should now have synchronized LFO and envelope clocks.

**Comment by fabiangreffrath** [comment]
Great, thank you!

**Review by rfomin — APPROVED** [review]
Looks good to me, thanks!

**Review by fabiangreffrath — APPROVED** [review]
Looks good to me, too. Thank you!

**Pull request merged**